ext/page_log/palm/palm.c

```diff
@@ -524,11 +525,20 @@ palm_add_reference(WT_PAGE_LOG *page_log)
     palm = (PALM *)page_log;
 
     /*
-     * Missing reference or overflow?
+     * Need to use CAS operation to avoid race conditions. The reason for not using fetch_add is
+     * that we need early failure.
      */
-    if (palm->reference_count == 0 || palm->reference_count + 1 == 0)
-        return (EINVAL);
-    ++palm->reference_count;
+    do {
+        unsigned int cur_count = atomic_load_explicit(&palm->reference_count, memory_order_relaxed);
+        /*
+         * Missing reference or overflow? Assume the first ref will not race the creation.
+         */
```

Comment:
My comment still applies, we should not change this comment.
```suggestion
        /* Missing reference or overflow? */
```